src/mpi/coll/allreduce/allreduce_init.c

```diff
@@ -0,0 +1,145 @@
+            break;
+        case MPIR_CVAR_IALLREDUCE_INTRA_ALGORITHM_gentran_recexch_multiple_buffer:
+            mpi_errno =
+                MPIR_TSP_Iallreduce_sched_intra_recexch(sendbuf, recvbuf, count, datatype, op,
+                                                        comm_ptr,
+                                                        MPIR_IALLREDUCE_RECEXCH_TYPE_MULTIPLE_BUFFER,
+                                                        MPIR_CVAR_IALLREDUCE_RECEXCH_KVAL, sched);
+            break;
+        case MPIR_CVAR_IALLREDUCE_INTRA_ALGORITHM_gentran_tree:
+            /*Only knomial_1 tree supports non-commutative operations */
+            MPII_COLLECTIVE_FALLBACK_CHECK(comm_ptr->rank, is_commutative ||
+                                           MPIR_Iallreduce_tree_type == MPIR_TREE_TYPE_KNOMIAL_1,
+                                           mpi_errno,
+                                           "Pallreduce gentran_tree cannot be applied.\n");
+            mpi_errno =
+                MPIR_TSP_Iallreduce_sched_intra_tree(sendbuf, recvbuf, count, datatype, op,
+                                                     comm_ptr, MPIR_Iallreduce_tree_type,
+                                                     MPIR_CVAR_IALLREDUCE_TREE_KVAL,
+                                                     MPIR_CVAR_IALLREDUCE_TREE_PIPELINE_CHUNK_SIZE,
+                                                     MPIR_CVAR_IALLREDUCE_TREE_BUFFER_PER_CHILD,
+                                                     sched);
+            break;
+        case MPIR_CVAR_IALLREDUCE_INTRA_ALGORITHM_gentran_ring:
+            MPII_COLLECTIVE_FALLBACK_CHECK(comm_ptr->rank, is_commutative, mpi_errno,
+                                           "Pallreduce gentran_ring cannot be applied.\n");
+            mpi_errno =
+                MPIR_TSP_Iallreduce_sched_intra_ring(sendbuf, recvbuf, count, datatype,
+                                                     op, comm_ptr, sched);
+            break;
+        case MPIR_CVAR_IALLREDUCE_INTRA_ALGORITHM_gentran_recexch_reduce_scatter_recexch_allgatherv:
+            MPII_COLLECTIVE_FALLBACK_CHECK(comm_ptr->rank, is_commutative &&
+                                           count >= nranks, mpi_errno,
+                                           "Pallreduce gentran_recexch_reduce_scatter_recexch_allgatherv cannot be applied.\n");
+            /*  This algorithm will work for commutative operations and if the count is
+             * bigger than total number of ranks. If it not commutative or if the count < nranks,
+             * MPIR_Iallreduce_sched algorithm will be run */
+            mpi_errno =
+                MPIR_TSP_Iallreduce_sched_intra_recexch_reduce_scatter_recexch_allgatherv
+                (sendbuf, recvbuf, count, datatype, op, comm_ptr,
+                 MPIR_CVAR_IALLREDUCE_RECEXCH_KVAL, sched);
+            break;
```

Comment:
It really bothers me that we are duplicating these codes. If you added the `sched` support, then I think you can refactor the code and avoid this duplication, right?

Reply:
I agree. It is getting worse when non-gentran algorithms are added.